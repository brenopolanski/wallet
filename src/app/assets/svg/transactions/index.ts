import { ReactComponent as DelegateRegistration } from "app/assets/svg/transactions/delegate-registration.svg";
import { ReactComponent as DelegateResignation } from "app/assets/svg/transactions/delegate-resignation.svg";
import { ReactComponent as Ipfs } from "app/assets/svg/transactions/ipfs.svg";
import { ReactComponent as Magistrate } from "app/assets/svg/transactions/magistrate.svg";
import { ReactComponent as Multipayment } from "app/assets/svg/transactions/multipayment.svg";
import { ReactComponent as Multisignature } from "app/assets/svg/transactions/multisignature.svg";
import { ReactComponent as Received } from "app/assets/svg/transactions/received.svg";
import { ReactComponent as Return } from "app/assets/svg/transactions/return.svg";
import { ReactComponent as SecondSignature } from "app/assets/svg/transactions/second-signature.svg";
import { ReactComponent as Sent } from "app/assets/svg/transactions/sent.svg";
import { ReactComponent as Timelock } from "app/assets/svg/transactions/timelock.svg";
import { ReactComponent as Transfer } from "app/assets/svg/transactions/transfer.svg";
import { ReactComponent as UnlockToken } from "app/assets/svg/transactions/unlock-token.svg";
import { ReactComponent as Unvote } from "app/assets/svg/transactions/unvote.svg";
import { ReactComponent as Vote } from "app/assets/svg/transactions/vote.svg";
import { ReactComponent as VoteCombination } from "app/assets/svg/transactions/vote-combination.svg";

export const TransactionIcons: any = {
	DelegateRegistration,
	DelegateResignation,
	Ipfs,
	Magistrate,
	Multipayment,
	Multisignature,
	Received,
	Return,
	SecondSignature,
	Sent,
	Timelock,
	Transfer,
	UnlockToken,
	Unvote,
	Vote,
	VoteCombination,
};
