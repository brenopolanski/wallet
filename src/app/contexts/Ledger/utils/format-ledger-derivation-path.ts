import { LedgerDerivationScheme } from "app/contexts/Ledger/contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
