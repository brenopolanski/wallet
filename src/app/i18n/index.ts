import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// Helpers
import { buildTranslations } from "./helpers";

const defaultNS = "translation";

const resources = {
	en: {
		translation: buildTranslations(),
	},
};

// eslint-disable-next-line
i18n.use(initReactI18next).init({
	defaultNS,
	lng: "en",
	ns: [defaultNS],
	resources,
});

export { defaultNS, i18n, resources };
