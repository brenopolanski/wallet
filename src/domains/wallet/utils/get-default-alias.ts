import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let sameCoinWalletsCount = sameCoinWallets.length || 1;

	while (profile.wallets().findByAlias(makeAlias(sameCoinWalletsCount))) {
		sameCoinWalletsCount++;
	}

	return makeAlias(sameCoinWalletsCount);
};
