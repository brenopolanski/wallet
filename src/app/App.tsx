// import { TRX } from "@payvo/sdk-trx";
// import { XLM } from "@payvo/sdk-xlm";
// import { XRP } from "@payvo/sdk-xrp";
// import { ZIL } from "@payvo/sdk-zil";
import "focus-visible";

import LedgerTransportNodeHID from "@ledgerhq/hw-transport-node-hid-singleton";
// import { LUNA } from "@payvo/sdk-luna";
// import { NANO } from "@payvo/sdk-nano";
// import { NEO } from "@payvo/sdk-neo";
import { Environment } from "@payvo/profiles";
// import { ADA } from "@payvo/sdk-ada";
import { ARK } from "@payvo/sdk-ark";
// import { ATOM } from "@payvo/sdk-atom";
// import { AVAX } from "@payvo/sdk-avax";
// import { BTC } from "@payvo/sdk-btc";
// import { DOT } from "@payvo/sdk-dot";
// import { EGLD } from "@payvo/sdk-egld";
// import { ETH } from "@payvo/sdk-eth";
import { LSK } from "@payvo/sdk-lsk";
import { Offline } from "domains/error/pages";
import { ExchangeProvider } from "domains/exchange/contexts/Exchange";
import { Splash } from "domains/splash/pages";
import React, { useLayoutEffect, useState } from "react";
import { useErrorHandler } from "react-error-boundary";
import { I18nextProvider, useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import { middlewares, RouterView, routes } from "router";
import { StubStorage } from "tests/mocks";
import { setThemeSource, shouldUseDarkColors } from "utils/electron-utils";
import { bootEnvWithProfileFixtures, isE2E, isUnit } from "utils/test-helpers";

import { SyncErrorMessage } from "./components/ProfileSyncStatusMessage";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, useEnvironmentContext } from "./contexts";
import { useDeeplink, useNetworkStatus, useProfileSynchronizer } from "./hooks";
import { i18n as index18n } from "./i18n";
import { PluginProviders } from "./PluginProviders";
import { SentryProvider } from "./sentry/SentryProvider";
import { SentryRouterWrapper } from "./sentry/SentryRouterWrapper";
import { httpClient, toasts } from "./services";

const RouteWrappers = ({ children }: { children: React.ReactNode }) => (
	<SentryRouterWrapper>{children}</SentryRouterWrapper>
);

const Main = () => {
	const [showSplash, setShowSplash] = useState(true);
	const { env } = useEnvironmentContext();
	const isOnline = useNetworkStatus();
	const history = useHistory();
	const { t } = useTranslation();

	useProfileSynchronizer({
		onProfileRestoreError: () => history.push("/"),
		onProfileSignOut: () => {
			toasts.dismiss();
		},
		onProfileSyncComplete: async () => {
			await toasts.dismiss();
			toasts.success(t("COMMON.PROFILE_SYNC_COMPLETED"));
		},
		onProfileSyncError: async (failedNetworkNames, retryProfileSync) => {
			await toasts.dismiss();

			toasts.warning(
				<SyncErrorMessage
					failedNetworkNames={failedNetworkNames}
					onRetry={async () => {
						await toasts.dismiss();
						retryProfileSync();
					}}
				/>,
			);
		},
		onProfileSyncStart: () => {
			toasts.warning(t("COMMON.PROFILE_SYNC_STARTED"), { autoClose: false });
		},
	});

	useDeeplink();

	useLayoutEffect(() => {
		setThemeSource("system");

		document.body.classList.remove(`theme-${shouldUseDarkColors() ? "light" : "dark"}`);
		document.body.classList.add(`theme-${shouldUseDarkColors() ? "dark" : "light"}`);
	}, []);

	const handleError = useErrorHandler();

	useLayoutEffect(() => {
		const boot = async () => {
			try {
				/* istanbul ignore next */
				if (isE2E() || isUnit()) {
					await bootEnvWithProfileFixtures({ env, shouldRestoreDefaultProfile: isUnit() });

					setShowSplash(false);
					return;
				}

				/* istanbul ignore next */
				await env.verify();
				/* istanbul ignore next */
				await env.boot();
			} catch (error) {
				handleError(error);
			}

			setShowSplash(false);
		};

		boot();
	}, [env, handleError]);

	const renderContent = () => {
		if (showSplash) {
			return <Splash />;
		}

		if (!isOnline) {
			return <Offline />;
		}

		return <RouterView routes={routes} middlewares={middlewares} wrapper={RouteWrappers} />;
	};

	return (
		<main data-testid="Main">
			<ToastContainer closeOnClick={false} newestOnTop />

			{renderContent()}
		</main>
	);
};

export const App = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */

	/* istanbul ignore next */
	const storage = isE2E() || isUnit() ? new StubStorage() : "indexeddb";

	const [environment] = useState(
		() =>
			new Environment({
				coins: {
					// ADA,
					ARK,
					// ATOM,
					// AVAX,
					// BTC,
					// DOT,
					// ETH,
					// EGLD,
					LSK,
					// NEO,
					// NANO,
					// LUNA,
					// TRX,
					// XLM,
					// XRP,
					// ZIL,
				},
				httpClient,
				storage,
			}),
	);

	return (
		<I18nextProvider i18n={index18n}>
			<EnvironmentProvider env={environment}>
				<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
					<SentryProvider>
						<LedgerProvider transport={LedgerTransportNodeHID}>
							<PluginProviders>
								<ExchangeProvider>
									<Main />
								</ExchangeProvider>
							</PluginProviders>
						</LedgerProvider>
					</SentryProvider>
				</ConfigurationProvider>
			</EnvironmentProvider>
		</I18nextProvider>
	);
};
