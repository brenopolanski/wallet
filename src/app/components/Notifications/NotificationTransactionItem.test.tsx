import { Contracts, DTO } from "@payvo/profiles";
import { httpClient } from "app/services";
import nock from "nock";
import React from "react";
import { TransactionFixture } from "tests/fixtures/transactions";
import { env, fireEvent, getDefaultProfileId, render, waitFor } from "utils/testing-library";

import { NotificationTransactionItem } from "./NotificationTransactionItem";
const NotificationTransactionsFixtures = require("tests/fixtures/coins/ark/devnet/notification-transactions.json");
const TransactionsFixture = require("tests/fixtures/coins/ark/devnet/transactions.json");

let profile: Contracts.IProfile;
let notificationTransaction: DTO.ExtendedConfirmedTransactionData;

describe("Notifications", () => {
	beforeEach(async () => {
		httpClient.clearCache();

		nock("https://ark-test.payvo.com").get("/api/transactions").query(true).reply(200, {
			data: NotificationTransactionsFixtures.data,
			meta: TransactionsFixture.meta,
		});

		profile = env.profiles().findById(getDefaultProfileId());

		await env.profiles().restore(profile);
		await profile.sync();
		await profile.notifications().transactions().sync();
		notificationTransaction = profile.notifications().transactions().transaction(TransactionFixture.id());
	});

	it("should render notification item", async () => {
		const { container, getAllByTestId } = render(
			<table>
				<tbody>
					<NotificationTransactionItem transaction={notificationTransaction!} profile={profile} />
				</tbody>
			</table>,
		);
		await waitFor(() => expect(getAllByTestId("TransactionRowMode")).toHaveLength(1));

		expect(container).toMatchSnapshot();
	});

	it("should emit onVisibilityChange event", async () => {
		const onVisibilityChange = jest.fn();

		const { getAllByTestId } = render(
			<table>
				<tbody>
					<NotificationTransactionItem
						transaction={notificationTransaction!}
						profile={profile}
						onVisibilityChange={onVisibilityChange}
					/>
				</tbody>
			</table>,
		);
		await waitFor(() => expect(getAllByTestId("TransactionRowMode")).toHaveLength(1));
		await waitFor(() => expect(onVisibilityChange).toHaveBeenCalledWith(expect.any(Boolean)));
	});

	it("should emit events onTransactionClick", async () => {
		const onTransactionClick = jest.fn();

		const { getAllByTestId, getByTestId } = render(
			<table>
				<tbody>
					<NotificationTransactionItem
						transaction={notificationTransaction!}
						profile={profile}
						onTransactionClick={onTransactionClick}
					/>
				</tbody>
			</table>,
		);
		await waitFor(() => expect(getAllByTestId("TransactionRowMode")).toHaveLength(1));

		fireEvent.click(getByTestId("TransactionRowMode"));

		await waitFor(() => expect(onTransactionClick).toHaveBeenCalledWith(notificationTransaction));
	});
});
