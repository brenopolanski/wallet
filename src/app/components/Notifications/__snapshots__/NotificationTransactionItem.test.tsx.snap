// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should render notification item 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="TableRow__TableRowStyled-sc-y9j4av-0 gwDXhN group"
        data-testid="TableRow"
      >
        <td
          class="w-3/5"
        >
          <div
            class="TableCell__TableCellInnerWrapper-sc-v39gua-0 jtRNyC flex space-x-3"
          >
            <div
              class="flex items-center space-x-2"
              data-testid="TransactionRowMode"
            >
              <span
                class="w-5 h-5 flex items-center border-0 border-theme-success-200 text-theme-success-600 dark:border-theme-success-600"
              >
                <div
                  class="sc-bdnxRM NjstW"
                  height="20"
                  width="20"
                >
                  <svg>
                    received.svg
                  </svg>
                </div>
              </span>
              <div
                class="Avatar__AvatarWrapper-sc-18vw8a4-0 fIpXvk ring-theme-background bg-theme-background group-hover:ring-theme-secondary-100 group-hover:bg-theme-secondary-100 dark:group-hover:ring-black dark:group-hover:bg-black flex-shrink-0"
                data-testid="Avatar"
              >
                <div
                  class="w-full h-full inline-flex items-center justify-center overflow-hidden align-middle rounded-full"
                >
                  <img
                    alt="DK1f5frUPTNqhFxoyDUHwN7VC2B4Uzhpmy"
                    src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"picasso\\" width=\\"100\\" height=\\"100\\" viewBox=\\"0 0 100 100\\"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(255, 193, 7)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"40\\" cx=\\"40\\" cy=\\"60\\" fill=\\"rgb(205, 220, 57)\\"/><circle r=\\"45\\" cx=\\"70\\" cy=\\"50\\" fill=\\"rgb(33, 150, 243)\\"/><circle r=\\"35\\" cx=\\"0\\" cy=\\"40\\" fill=\\"rgb(156, 39, 176)\\"/></svg>"
                    title="DK1f5frUPTNqhFxoyDUHwN7VC2B4Uzhpmy"
                  />
                </div>
              </div>
            </div>
            <div
              class="w-20 flex-1"
            >
              <div
                class="flex overflow-hidden flex-grow items-center space-x-2 whitespace-nowrap no-ligatures"
              >
                <span
                  class="inline-flex overflow-hidden text-theme-text font-semibold text-base"
                  data-testid="Address__address"
                >
                  D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
                </span>
              </div>
            </div>
          </div>
        </td>
        <td>
          <div
            class="TableCell__TableCellInnerWrapper-sc-v39gua-0 kXCVVq justify-end"
          >
            <div
              class="Label-sc-4akkzw-0 bdvTBm"
              color="success"
            >
              <div
                class="flex space-x-1 items-stretch"
              >
                <span
                  class="whitespace-nowrap"
                  data-testid="Amount"
                >
                  + 20 DARK
                </span>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
